File: php-backend/src/PlayerData/UseCase/PlayerDataUseCase.php
<?php


namespace TheScore\PlayerData\UseCase;


use TheScore\PlayerData\Repository\PlayerDataRepositoryInJsonFile;

class PlayerDataUseCase
{

    private PlayerDataRepositoryInJsonFile $repository;

    public function __construct(PlayerDataRepositoryInJsonFile $repository)
    {
        $this->repository = $repository;
    }

    public function getRecordsAsJson(string $nameToFilter, array $order, int $page, int $recordsPerPage): string
    {
        $allRecords = $this->repository->getRecords($nameToFilter, $order, $page, $recordsPerPage);
        return json_encode(array_values($allRecords), JSON_PRETTY_PRINT);
    }

    public function getRecordsAsCsv(mixed $nameToFilter, array $order, int $page, int $recordsPerPage): string
    {
        $allRecords = $this->repository->getRecords($nameToFilter, $order, $page, $recordsPerPage);
        $csv = '';
        foreach ($allRecords as $record) {
            if (empty($csv)) {
                $csvHeader = array_keys($record);
                $csv .= implode(',', $csvHeader) . PHP_EOL;
            }

            $csvValues = array_values($record);
            $csv .= implode(',', $csvValues) . PHP_EOL;
        }
        return $csv;
    }
}

File: elixir-backend/lib/player_data/use_case/player_data_use_case.ex
defmodule TheRush.PlayerData.UseCase.PlayerDataUseCase do

  def getRecordsAsJson(repository, filename, playerName, page, recordsPerPage, order) do
    repository.getRecords(filename, playerName, page, recordsPerPage, order)
    |>Jason.encode!
  end

  def getRecordsAsCsv(repository, filename, playerName, page, recordsPerPage, order) do
    records = repository.getRecords(filename, playerName, page, recordsPerPage, order)

    records
    |> Enum.at(0)
    |> Enum.to_list
    |> Enum.map_join(",", fn ({header, _}) -> header end)
    |> Kernel.<>("\n")
    |> Kernel.<>(Enum.map_join(records, "\n", &mapToCsvFormat(&1)))
    |> Kernel.<>("\n")
  end

  defp mapToCsvFormat(map) do
    map
    |> Enum.to_list
    |> Enum.map_join(",", fn ({_, value}) -> value end)
  end
end
